### ISSUE TITLE
False positive of "unused_braces" when combined with Mutex

### ISSUE THREAD

**Issue by kindabit** [rust-lang/rust#143562]
### Code

```Rust
use std::sync::Mutex;

struct Test {
    a: i32,
    b: i32,
}

struct VecVec {
    v: Vec<i32>,
}

impl VecVec {
    pub fn push(mut self, value: i32) -> Self {
        self.v.push(value);
        self
    }
}

pub fn main() {
    let test = Mutex::new(Test {
        a: 12,
        b: 34,
    });
    
    let vec_vec = VecVec { v: Vec::new() }
        .push({ test.lock().unwrap().a })  // the program will hang if this pair of braces is removed
        .push({ test.lock().unwrap().b });
    
    println!("len: {}", vec_vec.v.len());
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
warning: unnecessary braces around method argument
  --> src/lib.rs:27:15
   |
27 |         .push({ test.lock().unwrap().b });
   |               ^^                      ^^
   |
   = note: `#[warn(unused_braces)]` on by default
help: remove these braces
   |
27 -         .push({ test.lock().unwrap().b });
27 +         .push(test.lock().unwrap().b);
   |

warning: unnecessary braces around method argument
  --> src/lib.rs:26:15
   |
26 |         .push({ test.lock().unwrap().a })  // the program will hang if this pair of braces is removed
   |               ^^                      ^^
   |
help: remove these braces
   |
26 -         .push({ test.lock().unwrap().a })  // the program will hang if this pair of braces is removed
26 +         .push(test.lock().unwrap().a)  // the program will hang if this pair of braces is removed
   |

warning: `playground` (lib) generated 2 warnings (run `cargo fix --lib -p playground` to apply 2 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.57s
```

### Desired output

```Shell
The compiler should detect the side effects and not treat these braces as unnecessary.
```

### Rationale and extra context

[Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=dc9447e2ff8967236532bf2750216079)

### Other cases

```Rust

```

### Rust Version

```Shell
This is the output of `rustc --version --verbose` on my computer:

rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-pc-windows-msvc
release: 1.88.0
LLVM version: 20.1.5

This issue could also be reproduced on rust playground.
```

**Comment by dream-dasher** [comment]
Another false-positive `unused_braces ([Rustc lint](https://doc.rust-lang.org/stable/nightly-rustc/rustc_lint/unused/static.UNUSED_BRACES.html)):


<img width="765" height="89" alt="Image" src="https://github.com/user-attachments/assets/03e06a38-9bd5-40b1-ba41-a64130333a89" />

```rust
impl std::fmt::Debug for ErrWrapper {
       fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { write!(f, "{}", self) }
}
```

I've seen this before as well: whenever a function is placed on a single line the function body `{...}` trigger the _unused_braces_ lint, though they are (as far as I can tell) quite necessary.

**Comment by kindabit** [comment]
> Another false-positive `unused_braces ([Rustc lint](https://doc.rust-lang.org/stable/nightly-rustc/rustc_lint/unused/static.UNUSED_BRACES.html)):
> <img alt="Image" width="765" height="89" src="https://private-user-images.githubusercontent.com/33399972/470926203-03e06a38-9bd5-40b1-ba41-a64130333a89.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM2NzU2NDQsIm5iZiI6MTc1MzY3NTM0NCwicGF0aCI6Ii8zMzM5OTk3Mi80NzA5MjYyMDMtMDNlMDZhMzgtOWJkNS00MGIxLWJhNDEtYTY0MTMwMzMzYTg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI4VDA0MDIyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlYjUyN2Q1NzQyOTY2OTMxY2EzNzE2ZWVjNzA2MTIyYTFhMjM3MDAzZmEyYjgyNzk5MmIyZWIxOWQ2NmMyZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CamArWYmUhBGfPCoAwJ_kX-7BaYOS8Zwhi_AKQ4wfF8">
> 
> impl std::fmt::Debug for ErrWrapper {
>        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { write!(f, "{}", self) }
> }
> 
> I've seen this before as well: whenever a function is placed on a single line the function body `{...}` trigger the _unused_braces_ lint, though they are (as far as I can tell) quite necessary.

Have you tried the latest stable version? It seems that the false positive you mentioned has been fixed in 1.88.0 (I'm using VSCode with rust-analyzer btw)

<img width="1012" height="380" alt="Image" src="https://github.com/user-attachments/assets/4732a306-3ce6-4fce-9452-077339dab2a8" />

**Comment by xizheyin** [comment]
This may require more efforts. I don't have time for the time being. Leave it to interested people. :)
